README.md
# NarrativeGraphPipeline

## This repo contains code for incoming paper.

## 1. Prerequisites
I recommend using **conda**. To install conda environment, run the following:

>

**NarrativeQA** dataset version used is from **huggingface**.

## 2. How to run
Run the following:

> python run.py

## 3. Components

### 3.1. Data reading
Read stories and scripts from files, initially preprocess and decompose them into paragraphs.


CHANGELOG.md
# CHANGELOG

### This file contains changes and modifications during developing.

## July 20 2021, 00:03

- Apply new PATH deployment strategy: Narrative_utils stores anthing and its path is in .env
- Change some field names

## June 28 2021, 23:03

- Add new Scheduled Sampling strategy
- Reformat some field names

## June 20 2021, 14:43

- Apply data _version 4_ for training: not removing stopwords, new HTML removing, start-end extraction
- Reduce **n_paras** to _5_, **seq_len_ans** to _15_, increase **seq_len_para** to _170_
- Reduce **n_nodes** and **n_edges** in Graph module respectively
- Officially use BertVocab instead of own vocab
- Use **parquet** data format instead **csv**
- Change name of fields in _dataset.py_: **ques** to **ques_ids**, **paras** to **context_ids**,...

## June 10 2021, 11:09

- Add metric calculation for valid
- Add module _Preprocess_
- Remove _torch.nn.Embedding_ layer and replace by _BertEmbedding_
- Train with _BertVocab_ instead own vocab (BertTokenizer uses its own to tokenize)
- Increase `seq_len_para` to **182** instead of _122_
- Format files

## May 30 2021, 14:40

- Huge improvements: Use **Pytorch Lightning** and _template_
- Add new attentive strategy for _Finegrain_
- Modify module _Memory_
- Apply _BeamSearch_ modified from _huggingface_ implementation
- Discard _GloVe_
- Replace class _Vocab_ by _modified BertTokenizer_

## May 20 2021, 22:27

- Change code folder name from 'modules' to 'src'
- Add new graph-based memory architecture _MemoryGraph_ for reasoning
- Reuse module _FineGrain_ to embed question and paras
- Change _Trainer_ and _CustomDataset_ to align with new updates

## May 17 2021, 21:24

- Modify `configs.py` and fit it with data _version 3.0_
- Modify module _NarrativePipeline_: fit with new field names
- Modify `utils.py` of NarrativePipeline to fit with new field names, _data version 3_,...

## May 9 2021, 23:39

- Reimplement _BeamSearch_: Replace _max_depth_ and _max_breadth_ by _beam_size_
- Add **Nucleus Sampling** technique to _BeamSearch_
- Disable metric calculating fucntions. Intend to use built-in.
- Modify some components in _Transformer_
- Add field `ans1_loss` to DataLoader during training
- Use _BertTokenizer_ and remove some methods in class _Vocab_

## May 6 2021, 22:44

- Reorganize _configs.py_
- Inferring in Transformer is now totally independent flow
- Temporarily remove calculating metrics, just generating predicted answers only
- Add 2 more fields `ans1_plain` and `ans1_plain` as creating iterator. Those are for inferring only.
- Replace user-generated vocab by BERT vocab for creating data for training/eval/test.

## May 2 2021, 21:18

- Reimplement _BeamSearch_ by simplifiyng many steps
- Switching context now occurs forcibly after training an epoch
- Process in module "DataReading" now runs on single process
- Function _build_vocab_ adds special tokens to vocab file

## Apr 25 2021, 16:18

- Reimplement function calculating _ROUGE-L_, _BLEU-1_...
- Implement _Beam Search_
- Implement new answer inferring module namely _TransDecoder_ ans use _Beam Search_
- Add feature logging metric calculation in testing phase

## Apr 21 2021, 15:07

- Fix errors in processing _movie_ context in _DataReading_ module
- Reconstruct _DataReading_ module
- Replace **LSTM-based** by **BERT-based** _Embedding layer_
- Change arguments of _PGD's forward_ to fit with new _Embedding layer_
- Modify training process's variables in model to fit with new _Embedding layer_
- Reorganize _GraphReasoning_ module into new folder
- Use trick longer answer to calculate loss and 2 answers to calculate metrics

## Apr 19 2021, 10:53

- Change and reconstruct repo to fit with _GCN_ setting

## Apr 6 2021, 14:45

- Apply parallel processing and shared memory to _DataReading_ module
- Back up processed context file => speed up processing time

## Apr 6, 2021, 14:45

- Change module _data_reading_ significantly:
  - Construct entire module into class, class _DataReading_
  - Define may to process story and movie script separatedly
  - Process context before process each entry => speed up processsing
  - Apply top-down golden paras selection instead of tree-based

## Mar 30, 2021, 16:35

- Change minor thing in _IAL_ to speed up training
- Fix some problems in _trigger_inference_

## Mar 28, 2021, 22:17

- Reduce vocab size to 10 times smaller: Use top 1000 most occured words from context and words from answer
- Change in accumulating loss (detaching it before accumulating)
- Contiguous and change dimensions at some points in _PGD_
- Change the way of calculating _Equation 2_, hence reduce time and memory prohibitively spent during training

## Mar 26, 2021, 20:35

- Reimplement _Dataset_ module
- Finish test a training batch

## Mar 25, 2021, 11:18

- Finish remaining modules for training, testing and evaluating.

## Mar 19, 2021, 18:05

- Finish implementing _PointerGeneratorDecoder_
- Add question tokens to global vocab of _PointerGeneratorDecoder_

## Mar 16, 2021, 12:54

- Finish module _FineGrain_, _CustomDataset_, _IAL_

## Mar1, 2021, 16:50

- Finish _FineGrain_ module with single question and paragraph as input
- Modify _config_ file

## Mar 8, 2021, 12:00

- Abandon idea of using trained Bert-based Paras Selector and back to idea proposed in paper _Simple and Effective_ - using TfIdf

## Mar 2, 2021

- Use _csv_ format instead of _pickle_ for data backup
- Finish implementing training and inferring module for _ParasSelection_

## Feb 21, 2022

- Use 'DataFrame' in _pandas_ to pickle data and for easily deal with 'Dataset' of _huggingface_
- Reformat the data to be backed up

## Feb 17, 2021

- Reorder resources of paths of backup file. From now, paths are stored in `configs.py`

## Feb 16 2021, 15:34 PM

- Finish 'data_reading' module.

## Feb 13 2021, 10:18 AM

- Add 'data_reading' module to read, initially preprocess and decompose documents into paragraphs

## Feb 3 2021, 11:00 AM

- Very first commit

configs.py
"""This file processes arguments and configs"""

import argparse, logging, os

import torch


###############################
# Read arguments from CLI
###############################
parser = argparse.ArgumentParser()

parser.add_argument("--batch", type=int, default=5)
parser.add_argument("--num_proc", type=int, default=4, help="number of processes")
parser.add_argument("--n_epochs", type=int, default=10)
parser.add_argument("--lr", type=float, default=5e-4, help="Learning rate")
parser.add_argument("--w_decay", type=float, default=0, help="Weight decay")
parser.add_argument("--task", type=str, default="train", help="train | infer")
parser.add_argument("--is_debug", type=bool, default=False, help="true | false")

# args = parser.parse_args()
args, _ = parser.parse_known_args()


###############################
# Add several necessary argument
###############################


# args.device     = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
args.device     = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
args.multi_gpus = torch.cuda.device_count() > 0


args.seq_len_ques       = 40  + 2   # The reason why sequence length of ques, contx and ans
args.seq_len_para       = 120 + 2   # plus 2 is for CLS and SEP token
args.seq_len_ans        = 40  + 2   # maximum answer length of dataset
args.n_paras            = 30
args.d_embd             = 200
args.d_hid              = 64
args.max_len_ans        = 12        # maximum inferring steps of decoder
args.min_count_PGD      = 10        # min occurences of word to be added to vocab of PointerGeneratorDecoder
args.d_vocab            = 32715     # Vocab size
args.dropout            = 0.15
args.n_layers           = 5

args.beam_size          = 20
args.n_gram_beam        = 10

args.trans_nheads       = 4
args.trans_nlayers      = 3


###############################
# Config path of backup files
###############################
PATH    = {
    'raw_data_dir'      : [
        "/Users/hoangle/Projects/VinAI/_data/NarrativeQA",
        "/home/tommy/Projects/_data/NarrativeQA",
        "/home/ubuntu/NarrativeQA",
        "/root/NarrativeQA"
    ],
    'bert'              : [
        "/root/bert-base-uncased",
        "/home/ubuntu/bert-base-uncased",
        "/Users/hoangle/Projects/VinAI/_pretrained/bert-base-uncased",
        "/home/tommy/Projects/_pretrained/BERT/bert-base-uncased"
    ],
    'glove_embd'        : [
        ".vector_cache/",
        "/Users/hoangle/Projects/VinAI/_pretrained/bert-base-uncased",
        "/home/tommy/Projects/_pretrained/GloVe"
    ],

    ## Paths associated with Data Reading
    'processed_contx'   : "backup/proc_contx/[ID].json",
    'dataset'           : "backup/[SPLIT]/data_[SHARD].csv",
    'vocab'             : "backup/vocab.txt",

    ## Paths associated with models
    'saved_model'       : "backup/model.pt",
    'saved_chkpoint'    : "backup/chkpoint.pth.tar",
    'prediction'        : "backup/predictions.json",
    'memory'            : "backup/memory.pt",

    'log'               : "run.log"
}

for key, val in PATH.items():
    if isinstance(val, list):
        is_found = False
        for path in val:
            if os.path.isdir(path):
                is_found  = True
                PATH[key] = path
                break
        if not is_found:
            raise FileNotFoundError(f"{key} not found")
###############################
# Config logging
###############################
if args.is_debug:
    logging.basicConfig(format='%(asctime)s: %(message)s', datefmt='%b-%d-%Y %H:%M:%S')
else:
    logging.basicConfig(filename=PATH['log'], filemode='a+', format='%(asctime)s: %(message)s', datefmt='%b-%d-%Y %H:%M:%S')
logging.getLogger().setLevel(logging.INFO)
